### PR TITLE
Add knn to classifier comparison 4002

### PR DESCRIPTION
**Pull request by kcelia**
refs #https://github.com/zama-ai/concrete-ml-internal/issues/4018
closes #https://github.com/zama-ai/concrete-ml-internal/issues/4002

### CODE REVIEW

File: src/concrete/ml/sklearn/base.py [hunk 1]
```diff
@@ -1982,7 +1983,7 @@ def scatter1d(x, v, indices):
         # 1. Pairwise_euclidiean distance
         distance_matrix = pairwise_euclidean_distance(q_X)
 
-        # The square root in the Euclidean distance calculation is not applied to speed up FHE
+        # The square root in the Euclidean distance calculation is not applied to speed up Fgggg
```
Comment (andrei-stoian-zama) [inline src/concrete/ml/sklearn/base.py:1986]: what happened here :smile: ?

File: src/concrete/ml/sklearn/base.py [hunk 2]
```diff
@@ -2034,7 +2035,7 @@ def predict(self, X: Data, fhe: Union[FheMode, str] = FheMode.DISABLE) -> numpy.
         X = check_array_and_assert(X)
 
         topk_labels = []
-        for query in X:
+        for query in tqdm(X):
```
Comment (andrei-stoian-zama) [inline src/concrete/ml/sklearn/base.py:2038]: let's avoid tqdm

File: docs/advanced_examples/utils/knn_comparison.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+from functools import partial
+
+from classifier_comparison_utils import make_classifier_comparison
+
+from concrete.ml.sklearn import KNeighborsClassifier
+
+if __name__ == "__main__":
+    knn_classifiers = [
```
Comment (andrei-stoian-zama) [inline docs/advanced_examples/utils/knn_comparison.py:8]: 2bit quantization seems too low, can you try with 3 bit ?

**Review by andrei-stoian-zama — CHANGES_REQUESTED** [review]
This is going in the right direction. Can you please:

- try 3 or even 4bit quantization?
- add a comment in the NB cell that shows the image saying something like "The KNeighborsClassifier is not yet compatible with jupyter notebooks. To reproduce the result below, run the script `knn_comparison.py` "
- see other comments

File: src/concrete/ml/sklearn/base.py (hunk 1 above)
Comment (kcelia) [reply src/concrete/ml/sklearn/base.py:1986]: I have nooo idea lol

File: src/concrete/ml/sklearn/base.py (hunk 2 above)
Comment (kcelia) [reply src/concrete/ml/sklearn/base.py:2038]: It was just for me, to have an idea about status of the execution.

File: docs/advanced_examples/utils/classifier_comparison_utils.py [hunk 4]
```diff
@@ -21,30 +21,35 @@
...
     )
+    # pylint: disable-next=no-member
     rng = np.random.RandomState(2)
     X += 2 * rng.uniform(size=X.shape)
     linearly_separable = (X, y)
 
     datasets = [
-        make_moons(n_samples=200, noise=0.3, random_state=0),
-        make_circles(n_samples=200, noise=0.2, factor=0.5, random_state=1),
+        make_moons(n_samples=n_samples, noise=0.2, random_state=0),
```
Comment (kcelia) [inline docs/advanced_examples/utils/classifier_comparison_utils.py:51]: I reduced the noise especially for the knn experiments. 
The dataset is small, so these perturbations lead to outliers that perturb both the concrete and sklearn models.

File: docs/advanced_examples/utils/KNeighborsClassifiers.png
Comment (fd0r) [inline docs/advanced_examples/utils/KNeighborsClassifiers.png:1]: Great figure!

**Review by fd0r — DISMISSED** [review]
Good figure!

File: script/make_utils/check_all_images_are_used.sh [hunk 5]
```diff
@@ -7,6 +7,24 @@ set -e
...
+    if [[ $img_name =~ \.(jpeg|svg|jpg|png)$ ]] && [ -n "$mentioned_in_specific_ipynb" ]; then
+      MENTIONED_IMG_FILES+=("$img_file")
+    fi
+done
+
+# Remove the image file paths mentioned in notebooks from IMAGE_FILES.
+for mentioned_img_filename in "${MENTIONED_IMG_FILES[@]}"; do
+    IMAGE_FILES=$(echo "$IMAGE_FILES" | grep -vF "$mentioned_img_filename")
+done
+
```
Comment (kcelia) [inline script/make_utils/check_all_images_are_used.sh:27]: I saved the KNN comparison plot in 'docs/advanced_examples/utils.'

'check_all_images_are_used.sh' reports that the image isn't used.

So, I added this code to ensure that images mentioned in any notebook should pass

**Review by andrei-stoian-zama — CHANGES_REQUESTED** [review]
```
K-nearest neighbors: the k-nearest neighbors (KNN) algorithm is the very first distance-based algorithm implemented in FHE. Currently, the Concrete KNN requires more computational resources and time compared to its predecessors. Therefore, it is recommended to employ it on small data-sets. In terms of performance, the decision boundaries look stair-like due to the quantization effect and are generated using the predict method, as the predict_proba method is currently unavailable. Overall, Concrete KNN aligns with its FP23 counterpart
```

- "the k-nearest neighbors (KNN) algorithm is the very first distance-based algorithm implemented in FHE" -> not true in general, there are FHE KNN implementations in literature. 

Maybe just stay "Nearest-neighbors classification in FHE is challenging and requires more computational resources than other classifier types. Therefore, it is recommended to employ it on small data-sets (up to tens of examples) with strong quantization (n_bits <= 4). "

- subtitle "Distance based classifiers" -> Nearest-neighbors classifiers

**Comment by kcelia** [comment]
>  "the k-nearest neighbors (KNN) algorithm is the very first distance-based algorithm implemented in FHE" -> not true in general, there are FHE KNN implementations in literature.


Maybe just stay "Nearest-neighbors classification in FHE is challenging and requires more computational resources than other classifier types. Therefore, it is recommended to employ it on small data-sets (up to tens of examples) with strong quantization (n_bits <= 4). "

File: script/make_utils/check_all_images_are_used.sh (hunk 5 above)
Comment (fd0r) [reply script/make_utils/check_all_images_are_used.sh:27]: Why not changes this in check_all_images_are_used.py instead?

**Review by andrei-stoian-zama — APPROVED** [review]

**Review by fd0r — COMMENTED** [review]
Not sure about the check_all_images_are_used modification.
What do you think about it @andrei-stoian-zama ?

**Pull request merged**